stout/range.h

```diff
@@ -0,0 +1,102 @@
+#pragma once
+
+#include <algorithm>
+
+#include "stout/stream.h"
+
+////////////////////////////////////////////////////////////////////////
+
+namespace stout {
+namespace eventuals {
+
+////////////////////////////////////////////////////////////////////////
+
+struct _Range {
+  template <typename K_, typename Arg_>
+  struct Continuation : public detail::TypeErasedStream {
+    // NOTE: explicit constructor because inheriting 'TypeErasedStream'.
+    Continuation(K_ k, int from, int to, int step)
+      : k_(std::move(k)),
+        from_(from),
+        to_(to),
+        step_(step) {}
+
+    void Start() {
+      k_.Start(*this);
+    }
+
+    template <typename... Args>
+    void Fail(Args&&... args) {
+      k_.Fail(std::forward<Args>(args)...);
+    }
+
+    void Stop() {
+      k_.Stop();
+    }
+
+    void Ended() {
+      k_.Ended();
+    }
+
+    void Next() override {
+      if (from_ == to_) {
+        k_.Ended();
+      } else {
+        int temp = from_;
+        from_ += step_;
+        k_.Body(temp);
+      }
+    }
+
+    void Done() override {
+      k_.Ended();
+    }
+
+    K_ k_;
+    int from_;
+    const int to_;
+    const int step_;
+  };
+
+  struct Composable {
+    template <typename>
+    using ValueFrom = int;
+
+    template <typename Arg, typename K>
+    auto k(K k) && {
+      return Continuation<K, Arg>{std::move(k), from_, to_, step_};
+    }
+
+    const int from_;
+    const int to_;
+    const int step_;
+  };
+};
+
+inline auto Range(int to) {
+  if (to >= 0) {
+    return _Range::Composable{0, to, 1};
+  } else {
+    return _Range::Composable{0, 0, 0};
+  }
```

Comment:
Right now, we essentially check for (and deal with) this scenario in three different ways in our three different `Range(...)` methods. For readability and make-sure-it-always-works-ness, it seems better to deal with invalid ranges only once: in `Next()`. If a range is found to be invalid, simply call `Ended()`.

Can you give that a try and see how it looks?